# Student performance 

## Life cycle of machine learning project
1. Understanding the problem statement
2. Data collection.
3. Data checks to perform .
4. Exploratory data analysis.
5. Data pre-processing .
6. Model Training .
7. Choose best model.

## 1. Problem statement
a. This project understand how the students performance (test scores) is affected by other variables such as Gender,Ethnicity,parental level of education ,Lunch and test preparation course.
## 2 Data collection
a. Data consist of 8 column and 1000 rows.
## 2.1 Import Data and required packages
import Pandas ,Numpy ,seaborn

In [23]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

## Import the csv data

In [2]:
df = pd.read_csv("stud.csv")

In [3]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [4]:
df.shape

(1000, 8)

## 2.2 Dataset information
1. Gender : sex of students -> (Male/Female)
2. Race/Ethnicity -> (Group A,B,C,D,E)
3. Parental level of education : parent's final education ->(bachlor's Degree, some college , master's degree, associations degree,high school)
4. lunch : Having lunch before test(standard or free/reduced)
5. test preparation course : complete or not complete before test.
6. math score
7. Reading score
8. writing score

## Data Checks to perform
1. check missing values
2. check duplicates
3. check data type
4. check the number of unique values of each column
5. check statistics of dataset.
6. check various categories present in the diffrent categorical column

### 3.1 check missing values

In [6]:
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

#### There are no missing values

### 3.2 Check Duplicates

In [8]:
df.duplicated().sum()

0

### There are no duplicates

### 3.3 Check the datatypes 

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


### 3.4 Checking the number of unique values of each column

In [11]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

### 3.5 Check the statistics of data set

In [12]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

### Insight
1. from above description of numerical data , all means are very close to each other - between 66 and 68.05;
2. All standard deviation are also close - between 14.6 and 15.19;
3. while there is minimum score 0 for math , for writing minimum is much higher = 10 and for reading minimum is higher =17m

### 3.7 Exploaratory data

In [15]:
print("categories in 'gender' variables :  ",end=" ")
print(df["gender"].unique())

print("categories in 'race_ethnicity' variables :  ",end=" ")
print(df["race_ethnicity"].unique())

print("categories in 'parental_level_of_education' variables :  ",end=" ")
print(df["parental_level_of_education"].unique())

print("categories in 'lunch' variables :  ",end=" ")
print(df["lunch"].unique())

print("categories in 'test_preparation_course' variables :  ",end=" ")
print(df["test_preparation_course"].unique())

print("categories in 'math_score' variables :  ",end=" ")
print(df["math_score"].unique())

print("categories in 'reading_score' variables :  ",end=" ")
print(df["reading_score"].unique())

print("categories in 'writing_score' variables :  ",end=" ")
print(df["writing_score"].unique())

categories in 'gender' variables :   ['female' 'male']
categories in 'race_ethnicity' variables :   ['group B' 'group C' 'group A' 'group D' 'group E']
categories in 'parental_level_of_education' variables :   ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
categories in 'lunch' variables :   ['standard' 'free/reduced']
categories in 'test_preparation_course' variables :   ['none' 'completed']
categories in 'math_score' variables :   [ 72  69  90  47  76  71  88  40  64  38  58  65  78  50  18  46  54  66
  44  74  73  67  70  62  63  56  97  81  75  57  55  53  59  82  77  33
  52   0  79  39  45  60  61  41  49  30  80  42  27  43  68  85  98  87
  51  99  84  91  83  89  22 100  96  94  48  35  34  86  92  37  28  24
  26  95  36  29  32  93  19  23   8]
categories in 'reading_score' variables :   [ 72  90  95  57  78  83  43  64  60  54  52  81  53  75  89  32  42  58
  69  73  71  74  70  65  87  56  61  84  55  44  41 

In [19]:
## define numerical and categorical columns

numeric_features = [feature for feature in df.columns if df[feature].dtype !='O']
categorical_features = [feature for feature in df.columns if df[feature].dtype =='O']

## Print columns
print('we have {} numerical feature : {}'.format(len(numeric_features),numeric_features))
print('\nwe have {} categorical feature : {}'.format(len(categorical_features),categorical_features))

we have 3 numerical feature : ['math_score', 'reading_score', 'writing_score']

we have 5 categorical feature : ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


### 3.8 Adding column for "Total Score" and "Average"

In [20]:
df['total_score'] = df['math_score'] + df["reading_score"] + df["writing_score"]
df['average'] = df['total_score']/3
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

In [21]:
reading_full = df[df["reading_score"]==100]['average'].count()
writting_full = df[df['writing_score']==100]['average'].count()
math_full = df[df["math_score"]==100]['average'].count()

print(f"Number of students with full marks in Maths : {math_full}")
print(f"Number of students with full marks in reading_full : {reading_full}")
print(f"Number of students with full marks in writting_full : {writting_full}")

Number of students with full marks in Maths : 7
Number of students with full marks in reading_full : 17
Number of students with full marks in writting_full : 14


In [22]:
reading_less_20 = df[df["reading_score"]<=20]['average'].count()
writting_fless_20  = df[df['writing_score']<=20]['average'].count()
math_less_20  = df[df["math_score"]<=20]['average'].count()

print(f"Number of students with full marks in eading_less_20 : {reading_less_20}")
print(f"Number of students with full marks in writting_fless_20 : {writting_fless_20}")
print(f"Number of students with full marks in math_less_20 : {math_less_20}")

Number of students with full marks in eading_less_20 : 1
Number of students with full marks in writting_fless_20 : 3
Number of students with full marks in math_less_20 : 4


### Insights
1. From above values we get students have performed worst in maths
2. Best performance in reading section.

## 4. Exploratory Data(visualization)
4.1 Visualize average score distribution to make some conclusion.
1. Histogram
2. Kernel distribution functions(KDE)

## 4.1.1 Histogram And KDE

In [27]:
fig,axs=plt.subplot(1,2,figsize=(15,7))
plt.subplot(121)
sns.histplot(data=df,x="average",bins=30,kde=True,color="g")
plt.subplot(122)
sns.histplot(data=df,x="average",kde=True,hue="gender")
plt.show()

TypeError: subplot() takes 1 or 3 positional arguments but 2 were given

<Figure size 640x480 with 0 Axes>